### PR TITLE
TST: add tests for has_z and add other common geometries

### PR DESCRIPTION
**Pull request by mwtoews**
This PR generally expands more testing with Z geometry types, and handling of empty geometries.

- A few more geometry types are added to `shapely.tests.common.all_types` including a few empty geometries
- Similarly, a few more geometries are also added to `all_types_z`
- Rename (e.g.) `test_to_wkt_3D` to `test_to_wkt_z`, since tests for GEOS 3.12 will add support for M geometries which are also "3D", and will be addressed in a separate PR.
- Add tests for `has_z`, which were missing.
- Several of the tests were re-written to use more `struct` representations of binary strings, rather than raw binary strings. This was done to make them easier to read and modify.
- Some parameterized tests had long auto-generated IDs, so these are improved with [`pytest.param(..., id="better_name")`](https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest-param)

This is split from #1808 which will be closed after some time. This PR is to be backported to `maint-2.0`.

### CODE REVIEW

File: shapely/tests/test_geometry.py [hunk 1]
```diff
@@ -200,8 +206,29 @@ def test_get_z_2d():
 
 @pytest.mark.parametrize("geom", all_types)
 def test_new_from_wkt(geom):
+    assert_geometries_not_equal = False
+    if geom.is_empty and shapely.get_num_geometries(geom) > 0:
```
Comment (caspervdw) [inline shapely/tests/test_geometry.py:210]: - This logic is repeated a few times. Maybe it is good to have a centralized function that checks if the empty-collection / GEOS version conditions are met?
- The `xfail` means that the test is executed, but expected to fail. Isn't the `skip` better in this case?

**Review by caspervdw — APPROVED** [review]
Solid work, some minor adjustments and then please merge!

**Comment by caspervdw** [comment]
> This PR is to be backported to maint-2.0.

I don't see a reasons; only bugfixes should be backported IMO

File: shapely/tests/test_constructive.py [hunk 2]
```diff
@@ -57,6 +57,13 @@
 @pytest.mark.parametrize("geometry", all_types)
 @pytest.mark.parametrize("func", CONSTRUCTIVE_NO_ARGS)
 def test_no_args_array(geometry, func):
+    if (
+        geometry.is_empty
+        and shapely.get_num_geometries(geometry) > 0
+        and func is shapely.node
+        and shapely.geos_version < (3, 8, 3)
+    ):
+        pytest.xfail("GEOS < 3.8.3 crashes with empty geometries")  # GEOS GH-601
```
[the remark this answers is not in the document]
Comment (mwtoews) [reply shapely/tests/test_constructive.py:66]: this test logic is here because `MIN_GEOS_VERSION = "3.8"` implies 3.8.0, which is a version still widely used, e.g. with [Ubuntu focal (20.04LTS)](https://packages.ubuntu.com/focal/libgeos-3.8.0) (also the main OS for my laptop).

File: shapely/tests/test_geometry.py [hunk 3]
```diff
@@ -139,12 +142,12 @@ def test_get_geometry_collection(geom):
 
 def test_get_type_id():
     actual = shapely.get_type_id(all_types).tolist()
-    assert actual == [0, 1, 2, 3, 4, 5, 6, 7, 7]
+    assert actual == [0, 1, 2, 3, 3, 4, 5, 6, 7, 7, 0, 1, 3, 4, 5, 6, 4, 5, 6, 7]
```
[the remark this answers is not in the document]
Comment (mwtoews) [reply shapely/tests/test_geometry.py:145]: good catch, and I've checked that they are properly documented to return -1

File: shapely/tests/test_geometry.py (hunk 1 above)
Comment (mwtoews) [reply shapely/tests/test_geometry.py:210]: - Logic for checking equal geometries that yield unequal result is generalized into a function
- Some of the `pytest.xfail` conditions are changed to `pytest.skip` (perhaps more should?)

**Comment by mwtoews** [comment]
> I don't see a reasons; only bugfixes should be backported IMO

Ok, that makes sense. I'd rather see more emphasis towards 2.1 anyways. Thanks for a thorough review!

----

It seems there are some recent failures with macos with geos 3.8.3 over the past week, all with the same error message:
> shapely.errors.GEOSException: IllegalArgumentException: Points of LinearRing do not form a closed linestring

Does anyone have a clue what this is?

**Comment by mwtoews** [comment]
The issue with macos-latest-x64 Py3.8 GEOS 3.8.3 is possibly resolved with a change in NumPy version. Previously Python 3.8 was paired with [numpy 1.16.2](https://pypi.org/project/numpy/1.16.2/#files), which doesn't have any built wheels and was thus building from source each time. I suspect this older version of NumPy had an issue with a relatively newer Python version.

Newer failures are related to https://github.com/libgeos/geos/pull/973, which I'll create a fix.
